xcm/xcm-builder/src/integration_tests.rs

```diff
@@ -0,0 +1,413 @@
+		assert_eq!(
+			mock::sent_xcm(),
+			vec![(
+				Parachain(PARA_ID).into(),
+				Xcm::ReserveAssetDeposit {
+					assets: vec![ConcreteFungible { id: Parent.into(), amount }],
+					effects: vec![
+						buy_execution(dest_weight),
+						DepositAsset {
+							assets: vec![All],
+							dest: Junction::AccountId32 {
+								network: NetworkId::Kusama,
+								id: ALICE.into()
+							}
+							.into()
+						},
+					]
+				}
+			)]
+		);
+	});
+}
+
+#[test]
+fn query_holding_works() {
+	use xcm::opaque::v0::prelude::*;
+	let para_acc: AccountId = ParaId::from(PARA_ID).into_account();
+	let balances = vec![(ALICE, INITIAL_BALANCE), (para_acc.clone(), INITIAL_BALANCE)];
+	kusama_like_with_balances(balances).execute_with(|| {
+		let other_para_id = 3000;
+		let amount = 10 * ExistentialDeposit::get();
+		let query_id = 1234;
+		let weight = 4 * BaseXcmWeight::get();
+		let r = XcmExecutor::<XcmConfig>::execute_xcm(
+			Parachain(PARA_ID).into(),
+			Xcm::WithdrawAsset {
+				assets: vec![ConcreteFungible { id: Null, amount }],
+				effects: vec![
+					buy_execution(weight),
+					Order::QueryHolding {
+						query_id,
+						dest: Parachain(PARA_ID).into(),
+						assets: vec![All],
+					},
+					Order::DepositAsset {
+						assets: vec![All],
+						dest: Parachain(other_para_id).into(),
+					},
+				],
+			},
+			weight,
+		);
+		assert_eq!(r, Outcome::Complete(weight));
+		let other_para_acc: AccountId = ParaId::from(other_para_id).into_account();
+		assert_eq!(Balances::free_balance(other_para_acc), amount);
+		assert_eq!(Balances::free_balance(para_acc), INITIAL_BALANCE - amount);
+		assert_eq!(
+			mock::sent_xcm(),
+			vec![(
+				Parachain(PARA_ID).into(),
+				Xcm::QueryResponse {
+					query_id,
+					response: Response::Assets(vec![ConcreteFungible {
+						id: Parent.into(),
+						amount
+					}])
+				}
+			)]
+		);
+	});
+}
+
+#[test]
+fn teleport_to_statemine_works() {
+	use xcm::opaque::v0::prelude::*;
+	let para_acc: AccountId = ParaId::from(PARA_ID).into_account();
+	let balances = vec![(ALICE, INITIAL_BALANCE), (para_acc.clone(), INITIAL_BALANCE)];
+	kusama_like_with_balances(balances).execute_with(|| {
+		let statemine_id = 1000;
+		let amount =  10 * ExistentialDeposit::get();
+		let teleport_effects = vec![
+			buy_execution(0),
+			Order::DepositAsset { assets: vec![All], dest: Parachain(PARA_ID).into() },
+		];
+		let weight = 3 * BaseXcmWeight::get();
+		let r = XcmExecutor::<XcmConfig>::execute_xcm(
+			Parachain(PARA_ID).into(),
+			Xcm::WithdrawAsset {
+				assets: vec![ConcreteFungible { id: Null, amount }],
+				effects: vec![
+					buy_execution(weight),
+					Order::InitiateTeleport {
+						assets: vec![All],
+						dest: Parachain(statemine_id).into(),
+						effects: teleport_effects.clone(),
+					},
+				],
+			},
+			weight,
+		);
+		assert_eq!(r, Outcome::Complete(weight));
+		assert_eq!(Balances::free_balance(para_acc), INITIAL_BALANCE - amount);
```

Comment:
could also check that statemine balance increased

Reply:
the local balance will not change because it's a teleport